dev/system-test/tracing.ts

```diff
@@ -0,0 +1,656 @@
+  // instance. (2) the user does not provide an instance and we find the
+  // global instance to use.
+  globalOpenTelemetry: boolean;
+
+  // Whether the Firestore instance should use gRPC or REST.
+  preferRest: boolean;
+}
+
+describe('Tracing Tests', () => {
+  let firestore: Firestore;
+  let tracerProvider: NodeTracerProvider;
+  let inMemorySpanExporter: InMemorySpanExporter;
+  let consoleSpanExporter: ConsoleSpanExporter;
+  let gcpTraceExporter: TraceExporter;
+
+  const spanIdToChildrenSpanIds = new Map<string, string[]>();
+  const spanIdToSpanData = new Map<string, ReadableSpan>();
+  let rootSpanIds: string[] = [];
+
+  function afterEachTest(config: TestConfig): Promise<void> {
+    spanIdToChildrenSpanIds.clear();
+    spanIdToSpanData.clear();
+    rootSpanIds = [];
+
+    return verifyInstance(firestore);
+  }
+
+  function getOpenTelemetryOptions(
+    config: TestConfig,
+    tracerProvider: TracerProvider
+  ): FirestoreOpenTelemetryOptions {
+    const options: FirestoreOpenTelemetryOptions = {
+      enableTracing: true,
+      tracerProvider: undefined,
+    };
+
+    // If we are *not* using a global OpenTelemetry instance, a TracerProvider
+    // must be passed to the Firestore SDK.
+    if (!config.globalOpenTelemetry) {
+      options.tracerProvider = tracerProvider;
+    }
+
+    return options;
+  }
+
+  function beforeEachTest(config: TestConfig) {
+    // Remove the global tracer provider in case anything was registered
+    // in order to avoid duplicate global tracers.
+    trace.disable();
+    context.disable();
+
+    // Set up a context manager.
+    const contextManager = new AsyncLocalStorageContextManager();
+    contextManager.enable();
+    context.setGlobalContextManager(contextManager);
+
+    console.log('in beforeEachTest');
+    // Create a new tracer and span processor for each test to make sure there
+    // are no overlaps when reading the results.
+    tracerProvider = new NodeTracerProvider({
+      sampler: new AlwaysOnSampler(),
+    });
+
+    inMemorySpanExporter = new InMemorySpanExporter();
+    consoleSpanExporter = new ConsoleSpanExporter();
+    gcpTraceExporter = new TraceExporter();
+
+    // Always add the console exporter for local debugging.
+    tracerProvider.addSpanProcessor(
+      new BatchSpanProcessor(consoleSpanExporter)
+    );
+
+    if (config.e2e) {
+      tracerProvider.addSpanProcessor(new BatchSpanProcessor(gcpTraceExporter));
+    } else {
+      tracerProvider.addSpanProcessor(
+        new BatchSpanProcessor(inMemorySpanExporter)
+      );
+    }
+
+    if (config.globalOpenTelemetry) {
+      trace.setGlobalTracerProvider(tracerProvider);
+    }
+
+    const settings: Settings = {
+      preferRest: config.preferRest,
+      openTelemetryOptions: getOpenTelemetryOptions(config, tracerProvider),
+    };
+    if (process.env.FIRESTORE_NAMED_DATABASE) {
+      settings.databaseId = process.env.FIRESTORE_NAMED_DATABASE;
```

Comment:
could this "settings.databaseId" be overwritten  by the if statement in line 188? 
knit: chain the 2 if statements together for readability.

Reply:
no, if line 183 is executed, then `!settings.databaseId` will be false and line 189 will not get executed.

Done.